src/vs/workbench/api/common/extHostCommands.ts

```diff
@@ -107,6 +111,15 @@ export class ExtHostCommands implements ExtHostCommandsShape {
 		});
 	}
 
+	onDidExecuteCommand(listener: (command: ICommandEvent) => any, thisArgs?: any, disposables?: IDisposable[]) {
+		this._proxy.$onDidExecuteCommand();
+		return this.onDidExecuteCommandEmitter(listener, thisArgs, disposables);
```

Comment:
Not sure what this is about? Isn't this just re-exposing the event as function? Above exists `onDidExecuteCommandEmitter` (which isn't an emitter but an event btw) which should be enough. Also, I don't understand why the `$onDidExecuteCommand` call is needed? The main side only needs one listener for main-side events which it then forwards to the ext host side.

Reply:
are you referring to the return statement?  I got a little confused about what I should be returning from `onDidExecuteCommand`.  Most of the events seem to return a `Disposable`, so I was just trying to copy patterns.  Probably way off there.

With the `$onDidExecuteCommand` call, my thinking was that I wouldn't register a subscriber to listen to events on the main-side until an extension actually asked for it by calling `onDidExecuteCommand`.  Does it make more sense to just have one listener always on the main side that forwards events over, even if no one is listening on the extension side?

Reply:
> With the $onDidExecuteCommand call, my thinking was that I wouldn't register a subscriber to listen to events on the main-side until an extension actually

That makes sense but now you add a listener on the main side for each listener on the extension side. When instantiating an `Emitter` you can provide options with callbacks that are invoked for the first and last listener. That's what you wanna use.

Reply:
Sorry, can you walk me through what you're thinking?  

Should the main-side listener be added only the first time an extension subscribes to the event, instead of having a one-to-one ratio with the extension side?  How would the first/last callbacks play into it?

Reply:
Yes, there is a single listener on the main side that forwards events to extension side. The extension side will have its own emitter/event to which extensions subscribe. That makes sure only one IPC message is send per event, not one per listener, and only if at least one listener exsits

Reply:
Moved the proxy calls into the `onFirstListenerDidAdd` and `onLastListenerRemove` hooks on the ext Emitter.

One thing:  I need to read up on Disposables a bit more.  I *think* I'm supposed to return `this.onDidExecuteCommand` here, so extensions get a `disposable` that they can push to `context.subscriptions`?  I'm not 100% here.  hopefully it's right.